pbsmetrics/go_metrics.go

```diff
@@ -121,6 +151,186 @@ func registerAdapterMetrics(registry metrics.Registry, adapterOrAccount string,
+
+	am, ok = me.accountMetrics[id]
+	if ok {
+		// Unlock and return as quickly as possible
+		return am
+	}
+	am = &accountMetrics{}
+	am.requestMeter = metrics.GetOrRegisterMeter(fmt.Sprintf("account.%s.requests", id), me.metricsRegistry)
+	am.bidsReceivedMeter = metrics.GetOrRegisterMeter(fmt.Sprintf("account.%s.bids_received", id), me.metricsRegistry)
+	am.priceHistogram = metrics.GetOrRegisterHistogram(fmt.Sprintf("account.%s.prices", id), me.metricsRegistry, metrics.NewExpDecaySample(1028, 0.015))
+	am.adapterMetrics = make(map[openrtb_ext.BidderName]*AdapterMetrics, len(me.exchanges))
+	for _, a := range me.exchanges {
+		am.adapterMetrics[a] = makeBlankAdapterMetrics(me.metricsRegistry, a)
+		registerAdapterMetrics(me.metricsRegistry, fmt.Sprintf("account.%s", id), string(a), am.adapterMetrics[a])
+	}
+
+	me.accountMetrics[id] = am
+
+	return am
+}
+
+// Implement the MetricsEngine interface
+
+// RecordRequest implements a part of the MetricsEngine interface
+func (me *Metrics) RecordRequest(labels Labels) {
+	me.RequestMeter.Mark(1)
+	if labels.Source == DemandApp {
+		me.AppRequestMeter.Mark(1)
+	} else {
+		if labels.Browser == BrowserSafari {
+			me.SafariRequestMeter.Mark(1)
+			if labels.CookieFlag == CookieFlagNo {
+				me.SafariNoCookieMeter.Mark(1)
+			}
+		}
+		if labels.CookieFlag == CookieFlagNo {
+			// NOTE: Old behavior was log me.AMPNoCookieMeter here for AMP requests.
+			// AMP is still new and OpenRTB does not do this, so changing to match
+			// OpenRTB endpoint
+			me.NoCookieMeter.Mark(1)
+		}
+	}
+	switch labels.RType {
+	case ReqTypeORTB2:
+		me.ORTBRequestMeter.Mark(1)
+	case ReqTypeAMP:
+		me.AmpRequestMeter.Mark(1)
+	}
+	if labels.RequestStatus == RequestStatusErr {
+		me.ErrorMeter.Mark(1)
+	}
+	// Handle the account metrics now.
+	am := me.getAccountMetrics(labels.PubID)
+	am.requestMeter.Mark(1)
+}
+
+// RecordRequestTime implements a part of the MetricsEngine interface. The calling code is responsible
+// for determining the call duration.
+func (me *Metrics) RecordRequestTime(labels Labels, length time.Duration) {
+	// Only record times for successful requests, as we don't have labels to screen out bad requests.
+	if labels.RequestStatus == RequestStatusOK {
+		me.RequestTimer.Update(length)
+	}
+}
+
+// RecordAdapterRequest implements a part of the MetricsEngine interface
+func (me *Metrics) RecordAdapterRequest(labels Labels) {
+	_, ok := me.AdapterMetrics[labels.Adapter]
+	if !ok {
+		glog.Errorf("Trying to run adapter metrics on %s: adapter metrics not found", string(labels.Adapter))
+		return
+	}
+	// Adapter metrics
+	am := me.AdapterMetrics[labels.Adapter]
+	am.RequestMeter.Mark(1)
+	// Account-Adapter metrics
+	aam := me.getAccountMetrics(labels.PubID).adapterMetrics[labels.Adapter]
+	aam.RequestMeter.Mark(1)
+
+	switch labels.RequestStatus {
+	case RequestStatusErr:
+		am.ErrorMeter.Mark(1)
+	case RequestStatusNoBid:
+		am.NoBidMeter.Mark(1)
+	case RequestStatusTimeout:
+		am.TimeoutMeter.Mark(1)
```

Comment:
Should we log an error in the `default` here? (and elsewhere)

Reply:
Given the previous behavior that we only log an error if we explicitly found an error, I felt this was the best match to the legacy metrics behavior. If there were no bugs, the two cases should be identical. Also if we add a default, then we need a case for `RequestStatusOk`

Reply:
ah, ok. yeah don't worry about it... not a big deal either way.